ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -88,14 +118,27 @@ def __call__(self, environ, start_response):
         tracer = trace.tracer()
         path_info = environ["PATH_INFO"] or "/"
 
-        with tracer.start_span(path_info, kind=trace.SpanKind.SERVER) as span:
+        span = tracer.create_span(path_info, kind=trace.SpanKind.SERVER)
+        span.start()
+        try:
             self._add_request_attributes(span, environ)
             start_response = self._create_start_response(span, start_response)
 
             iterable = self.wsgi(environ, start_response)
-            try:
-                for yielded in iterable:
-                    yield yielded
-            finally:
-                if hasattr(iterable, "close"):
-                    iterable.close()
+
+            # Put this in a subfunction to not delay the call to the wrapped
+            # WSGI application (instrumentation should change the application
+            # behavior as little as possible).
+            def iter_result():
+                try:
+                    for yielded in iterable:
+                        yield yielded
+                finally:
+                    if hasattr(iterable, "close"):
+                        iterable.close()
+                    span.end()
+
+            return iter_result()
+        except:  # noqa
+            span.end()
```

Comment:
This seems to be risky.
According to [the API](https://github.com/open-telemetry/opentelemetry-python/blob/8258927441478b6e793c8c2031b459203b481ba3/opentelemetry-api/src/opentelemetry/trace/__init__.py#L161), implementation could raise exception if `span.end()` is called twice.

If an exception happened right before `iter_result` returns (but after `span.end()` is called), we end up calling `end()` twice and kill the user app due to unwanted exception.

Reply:
We may want to change the API so that we don't raise on duplicate calls to `end`, which seems like a good idea regardless of this PR.

I've got an [open PR in specs](https://github.com/open-telemetry/opentelemetry-specification/pull/153) about suppressing errors in the API. The goal is to avoid exactly this kind of thing -- crashing the application because we're using the instrumentation layer wrong.

Reply:
Seems there is no way to do it right with the current API then.

Reply:
Probably leave a TODO comment for now.

Reply:
BTW, the current SDK only logs a warning, which is fine because having end called twice here should really be an edge case.

Reply:
For `end` in particular the spec already says that calling it multiple times is fine, see https://github.com/open-telemetry/opentelemetry-python/pull/154#issuecomment-534475659